### ISSUE TITLE
Vector UB usable size appears to be ~184KB (not 192KB) on Ascend910B1 / cann-9.0.0 — silent corruption when tile allocations exceed ~184KB

### ISSUE THREAD

**Issue by Hzfengsy** [PTO-ISA/pto-isa#170]
## Summary

On **Ascend910B1 / CANN `cann-9.0.0`** (target `a2a3`), a pure-Vector kernel whose `pto.alloc_tile` allocations reach a **188.8 KB** Vec high-water — well within the **192 KB** Vec UB — produces **NaN** in its output on device. The *same* kernel reduced to a **183.3 KB** high-water is correct. The corruption switches on **between 183.3 KB (clean) and 188.8 KB (NaN)** — i.e. right around **184 KB**. Nothing errors at compile in either case; the 188.8 KB case **silently corrupts** at runtime.

| case | Vec footprint | # buffers | device output |
| ---- | ------------- | --------- | ------------- |
| HD512 | **188.8 KB** / 192 KB | 81 | **NaN** (112 NaNs) |
| HD496 | **183.3 KB** / 192 KB | 81 | correct (0 NaN) |

Both are the *same* kernel with an identical op sequence; only the tile column size differs (→ footprint).

## CANN version (important)

The device reproduction runs on **cann-9.0.0** (active `ASCEND_HOME_PATH` / `LD_LIBRARY_PATH` / `PATH` all point at `.../Ascend/cann-9.0.0`, Version=9.0.0). We were told this Vec-UB reservation "was removed" — but **188.8 KB (< 192 KB) still corrupts on 9.0.0**, so it appears to still be in effect. (`cann-8.5.1` was used *only* for the in-core op-simulator, which needs a TL-capable `bisheng` for the camodel build; it does not run the device NaN.)

## Hypothesis

The real usable Vec UB is **~184 KB, not 192 KB** — i.e. ~8 KB at the top is reserved (we see `ReserveBufferOp` / `PTOInferValidatePipeInitPass` in PTOAS). A tile placed in that top region reads/writes garbage → NaN. The upstream (PyPTO) allocator's 192 KB budget does not subtract this, so it emits a `.pto` reaching 188.8 KB without any error.

## Evidence the NaN is memory corruption (not numerics, not a logic bug)

- **Localized**: every NaN is in the highest-address accumulator tiles (the `m_oi_nope` accumulator, NOPE half of the output); lower-address tiles are clean.
- **Footprint-causal**: drop the high-water below ~184 KB (smaller tiles, or set the inner pipeline to `stage=1`) → 0 NaN; same kernel otherwise.
- **Not a numerical artifact**: with inputs that keep the divisor `n_denom = li + exp(...)` strictly positive, the 188.8 KB case **still** NaNs → corrupted buffer data, not a division.
- **Data-dependent cells**: NaN cells vary with input seed (88 vs 112) — consistent with a high-address buffer overlapping a reserved/garbage region.

## Reproduction

Artifacts — both `.pto`, the per-buffer memory reports, and the repro scripts:
**https://gist.github.com/Hzfengsy/8f655587af225f73665a1e9a3112441a**

For PTOAS directly: compile `HD512_WRONG_merge_rope_pack.pto` and `HD496_CORRECT_merge_rope_pack.pto` for `a2a3` / `Ascend910B1` / cann-9.0.0, run, and check the output for NaN. HD512's top `alloc_tile` reaches addr ~188.6 KB (→ NaN); HD496 stays ≤ ~183.1 KB (→ clean). `diff` the two `.pto` — same structure, only tile sizes / `addr` values differ.

<details><summary>PyPTO-side standalone device harness — merge_h.py (HD env var sets the hidden dim / footprint)</summary>

```python
# Copyright (c) PyPTO Contributors.
# Standalone device harness: lifts the dsv4 `merge_rope_pack` scope out of
# decode_sparse_attn.py and runs it as its own kernel on device (-d 0) to check
# whether o_packed (its OWN output) contains NaN/Inf. Isolates whether the
# coloring corruption is intrinsic to the kernel.
import sys

sys.path.insert(0, "/home/syfeng/pypto-pipeline-stage-scope/python")
import pypto  # noqa: E402

assert "pipeline-stage-scope" in pypto.__file__, pypto.__file__
assert "pipeline-stage-scope" in pypto.pypto_core.__file__, pypto.pypto_core.__file__

sys.path.insert(0, "/home/syfeng/pypto/dsv4_kernels")
import pypto.language as pl  # noqa: E402
import decode_sparse_attn as M  # noqa: E402
import os

# Reference constants through M.* so they stay in sync with the kernel.
T = M.T
H = M.H
HEAD_DIM = int(os.environ.get('HD', M.HEAD_DIM))
NOPE_DIM = HEAD_DIM - M.ROPE_DIM
ROPE_DIM = M.ROPE_DIM
HALF_ROPE = M.HALF_ROPE
ROPE_H_TILE = M.ROPE_H_TILE
ROPE_SUBTILES = M.ROPE_SUBTILES
H_TILE = M.H_TILE
SPARSE_BLOCKS = M.SPARSE_BLOCKS
HEADS_PER_GROUP = M.HEADS_PER_GROUP
O_GROUPS = M.O_GROUPS
D = M.HEADS_PER_GROUP * HEAD_DIM
FUSE_ALIGNED_SINGLE_STORE = M.FUSE_ALIGNED_SINGLE_STORE

# T*(H//H_TILE)*SPARSE_BLOCKS*H_TILE = 128*4*5*16 = 40960
SPARSE_ROWS = T * (H // H_TILE) * SPARSE_BLOCKS * H_TILE
assert SPARSE_ROWS == 40960, SPARSE_ROWS


@pl.jit
def main(
    sparse_blk_mi: pl.Tensor[[40960, 1], pl.FP32],
    sparse_blk_li: pl.Tensor[[40960, 1], pl.FP32],
    sparse_blk_oi: pl.Tensor[[40960, HEAD_DIM], pl.FP32],
    attn_sink: pl.Tensor[[64], pl.FP32],
    freqs_cos: pl.Tensor[[128, 64], pl.BF16],
    freqs_sin: pl.Tensor[[128, 64], pl.BF16],
    o_packed: pl.Out[pl.Tensor[[1024, D], pl.BF16]],
):
    # ---- BEGIN verbatim lift of decode_sparse_attn.py lines 346..449 ----
    for m_t in pl.spmd(T, name_hint="merge_rope_pack"):
        m_token_base = m_t * (H // ROPE_H_TILE) * SPARSE_BLOCKS * ROPE_H_TILE
        # Inverse-RoPE interleave pattern (j^1 swap, j>>1 dup, [+1,-1,...] sign)
        # over the FULL ROPE_DIM at once (the standalone rope chunked into 32-col
        # tloads; here the rope tail is a single contiguous [ROPE_H_TILE, ROPE_DIM]
        # tile, so we rotate it in one pass -- no chunk-slicing). Column-only ->
        # build once per token-task, on ROPE_H_TILE rows.
        sp_ones = pl.full([ROPE_H_TILE, ROPE_DIM], dtype=pl.FP32, value=1.0)
        sp_col = pl.col_expand_mul(
            sp_ones, pl.cast(pl.arange(0, [1, ROPE_DIM], dtype=pl.INT32), target_type=pl.FP32)
        )
        sp_dup_f = pl.cast(
            pl.cast(pl.mul(sp_col, 0.5), target_type=pl.INT32, mode="trunc"), target_type=pl.FP32
        )
        sp_dup_idx = pl.cast(sp_dup_f, target_type=pl.INT32)  # j>>1
        sp_lane = pl.sub(sp_col, pl.mul(sp_dup_f, 2.0))  # j%2
        sp_swap_idx = pl.cast(pl.sub(pl.add(sp_col, 1.0), pl.mul(sp_lane, 2.0)), target_type=pl.INT32)  # j^1
        sp_sign = pl.neg(pl.sub(pl.mul(sp_lane, 2.0), 1.0))  # [+1,-1,...] (conjugate)

        for m_h_idx in pl.pipeline(H // ROPE_H_TILE, stage=2):
            m_h0 = m_h_idx * ROPE_H_TILE
            # sparse_blk_* is stored in H_TILE-row blocks (qk_pv layout). Map this
            # ROPE_H_TILE-row sub-tile to its stored head-tile m_ht and the half
            # m_half within that block; the base offsets into the block by
            # m_half * ROPE_H_TILE, and the sparse-block stride below stays H_TILE
            # (the storage stride), not ROPE_H_TILE.
            m_ht = m_h_idx // ROPE_SUBTILES
            m_half = m_h_idx % ROPE_SUBTILES
            m_blk_base = m_token_base + m_ht * SPARSE_BLOCKS * H_TILE + m_half * ROPE_H_TILE
            m_mi = sparse_blk_mi[m_blk_base : m_blk_base + ROPE_H_TILE, 0:1]
            m_li = sparse_blk_li[m_blk_base : m_blk_base + ROPE_H_TILE, 0:1]
            # Split the attention-output accumulator into NOPE + ROPE column
            # ranges, each loaded as its own contiguous tile (tensor slice ->
            # tload). PyPTO Vec ops need full contiguous operands, so the rope
            # tail must be a standalone tile, not a column-subview of a wide one.
            m_oi_nope = sparse_blk_oi[m_blk_base : m_blk_base + ROPE_H_TILE, 0:NOPE_DIM]
            m_oi_rope = sparse_blk_oi[m_blk_base : m_blk_base + ROPE_H_TILE, NOPE_DIM:HEAD_DIM]

            for m_sb in pl.pipeline(1, SPARSE_BLOCKS, stage=2):
                m_row = m_blk_base + m_sb * H_TILE
                m_cur_mi = sparse_blk_mi[m_row : m_row + ROPE_H_TILE, 0:1]
                m_cur_li = sparse_blk_li[m_row : m_row + ROPE_H_TILE, 0:1]
                m_cur_oi_nope = sparse_blk_oi[m_row : m_row + ROPE_H_TILE, 0:NOPE_DIM]
                m_cur_oi_rope = sparse_blk_oi[m_row : m_row + ROPE_H_TILE, NOPE_DIM:HEAD_DIM]
                m_mi_new = pl.maximum(m_mi, m_cur_mi)
                m_alpha = pl.exp(pl.sub(m_mi, m_mi_new))
                m_beta = pl.exp(pl.sub(m_cur_mi, m_mi_new))
                m_li = pl.add(pl.mul(m_alpha, m_li), pl.mul(m_beta, m_cur_li))
                m_oi_nope = pl.add(
                    pl.row_expand_mul(m_oi_nope, m_alpha), pl.row_expand_mul(m_cur_oi_nope, m_beta)
                )
                m_oi_rope = pl.add(
                    pl.row_expand_mul(m_oi_rope, m_alpha), pl.row_expand_mul(m_cur_oi_rope, m_beta)
                )
                m_mi = m_mi_new

            n_sink_bias = pl.reshape(attn_sink[m_h0 : m_h0 + ROPE_H_TILE], [ROPE_H_TILE, 1])
            n_sink_tile = pl.add(pl.sub(m_mi, m_mi), n_sink_bias)
            n_denom = pl.add(m_li, pl.exp(pl.sub(n_sink_tile, m_mi)))
            # NOPE half: normalize + BF16 (unchanged from the old NOPE store path).
            n_nope = pl.cast(pl.row_expand_div(m_oi_nope, n_denom), target_type=pl.BF16)
            # ROPE half: normalize, then BF16-round the rope INPUT (matches the old
            # attn_rope_stage BF16 round-trip that golden also does), inverse-RoPE
            # the contiguous [ROPE_H_TILE, ROPE_DIM] tile, BF16-round the output.
            #   out[j] = x[j]*cos_il[j] + x[j^1]*sign[j]*sin_il[j]
            n_rope_in = pl.cast(
                pl.cast(pl.row_expand_div(m_oi_rope, n_denom), target_type=pl.BF16), target_type=pl.FP32
            )
            r_cos = pl.cast(freqs_cos[m_t : m_t + 1, 0:HALF_ROPE], target_type=pl.FP32)
            r_sin = pl.cast(freqs_sin[m_t : m_t + 1, 0:HALF_ROPE], target_type=pl.FP32)
            r_cos_h = pl.col_expand_mul(pl.full([ROPE_H_TILE, HALF_ROPE], dtype=pl.FP32, value=1.0), r_cos)
            r_sin_h = pl.col_expand_mul(pl.full([ROPE_H_TILE, HALF_ROPE], dtype=pl.FP32, value=1.0), r_sin)
            r_cos_il = pl.gather(r_cos_h, dim=-1, index=sp_dup_idx)
            r_sin_il = pl.gather(r_sin_h, dim=-1, index=sp_dup_idx)
            r_swapped = pl.gather(n_rope_in, dim=-1, index=sp_swap_idx)
            r_rot = pl.add(pl.mul(n_rope_in, r_cos_il), pl.mul(pl.mul(r_swapped, sp_sign), r_sin_il))
            n_rope = pl.cast(r_rot, target_type=pl.BF16, mode="rint")

            # Write each head to o_packed. Two store modes (compile-time switch,
            # FUSE_ALIGNED_SINGLE_STORE) -- both bit-identical in intent, they only
            # differ in store granularity / alignment:
            #   - single aligned: join [16,448]+[16,64] -> [16,512] via pl.concat
            #     and emit one [1,HEAD_DIM]=1024B 512B-aligned store per head.
            #     Faster (VECTOR-bound) but blocked by pto-isa TCONCAT bug #168.
            #   - two-store: emit NOPE [1,448] and ROPE [1,64] separately (same as
            #     the original NOPE + rope_pack stores). Correct today.
            if FUSE_ALIGNED_SINGLE_STORE:
                n_full = pl.concat(n_nope, n_rope)
                for n_hi in pl.range(ROPE_H_TILE):
                    n_gh = m_h0 + n_hi
                    n_g = n_gh // HEADS_PER_GROUP
                    n_hh = n_gh - n_g * HEADS_PER_GROUP
                    n_pack_row = n_g * T + m_t
                    n_col = n_hh * HEAD_DIM
                    o_packed[n_pack_row : n_pack_row + 1, n_col : n_col + HEAD_DIM] = n_full[n_hi : n_hi + 1, 0:HEAD_DIM]
            else:
                for n_hi in pl.range(ROPE_H_TILE):
                    n_gh = m_h0 + n_hi
                    n_g = n_gh // HEADS_PER_GROUP
                    n_hh = n_gh - n_g * HEADS_PER_GROUP
                    n_pack_row = n_g * T + m_t
                    n_col = n_hh * HEAD_DIM
                    o_packed[n_pack_row : n_pack_row + 1, n_col : n_col + NOPE_DIM] = n_nope[n_hi : n_hi + 1, 0:NOPE_DIM]
                    o_packed[n_pack_row : n_pack_row + 1, n_col + NOPE_DIM : n_col + HEAD_DIM] = n_rope[n_hi : n_hi + 1, 0:ROPE_DIM]
    # ---- END verbatim lift ----
    return o_packed


def _build_specs():
    """7-entry TensorSpec list: 6 finite-random inputs + o_packed output."""
    import torch
    from golden import TensorSpec

    def seeded_uniform(shape, seed, dtype):
        def _init():
            gen = torch.Generator()
            gen.manual_seed(seed)
            # SMALL finite uniform[-0.5, 0.5] so online-softmax exp() cannot
            # overflow from data -- any NaN/Inf is therefore CORRUPTION, not math.
            return (torch.rand(*shape, generator=gen) - 0.5).to(dtype)

        return _init

    return [
        TensorSpec("sparse_blk_mi", [40960, 1], torch.float32, init_value=seeded_uniform((40960, 1), 11, torch.float32)),
        TensorSpec("sparse_blk_li", [40960, 1], torch.float32, init_value=seeded_uniform((40960, 1), 12, torch.float32)),
        TensorSpec("sparse_blk_oi", [40960, HEAD_DIM], torch.float32, init_value=seeded_uniform((40960, HEAD_DIM), 13, torch.float32)),
        # attn_sink = zeros (matches golden).
        TensorSpec("attn_sink", [64], torch.float32, init_value=0.0),
        TensorSpec("freqs_cos", [128, 64], torch.bfloat16, init_value=seeded_uniform((128, 64), 14, torch.bfloat16)),
        TensorSpec("freqs_sin", [128, 64], torch.bfloat16, init_value=seeded_uniform((128, 64), 15, torch.bfloat16)),
        TensorSpec("o_packed", [1024, D], torch.bfloat16, is_output=True),
    ]


def _golden_zero(values):
    """Trivial golden: zero-fill o_packed. _validate still prints the device
    output's illegal-values (NaN/Inf) line regardless of golden match."""
    import torch

    values["o_packed"] = torch.zeros((1024, D), dtype=torch.bfloat16)


def main_run(device: int = 0, platform: str = "a2a3") -> bool:
    from golden import ratio_allclose, run_jit

    specs = _build_specs()
    runtime_cfg = dict(platform=platform, device_id=device, enable_l2_swimlane=False)
    # Loose max_error_ratio so the ratio check does not abort before the NaN
    # report; the NaN/Inf hard check inside the comparator always runs first and
    # prints "illegal values in actual: NaN=.. Inf=.." for the DEVICE output.
    compare_fn = {"o_packed": ratio_allclose(atol=1e9, rtol=1e9, max_error_ratio=1.0)}
    result = run_jit(
        fn=main,
        specs=specs,
        golden_fn=_golden_zero,
        save_data=False,
        rtol=1e9,
        atol=1e9,
        compare_fn=compare_fn,
        runtime_cfg=runtime_cfg,
    )
    print(f"[merge_rope_standalone] passed={result.passed} error={result.error!r}")
    return result.passed


if __name__ == "__main__":
    import argparse

    ap = argparse.ArgumentParser()
    ap.add_argument("-d", "--device", type=int, default=0)
    ap.add_argument("-p", "--platform", type=str, default="a2a3")
    ap.add_argument("--compile-only", action="store_true")
    args = ap.parse_args()

    if args.compile_only:
        from golden import run_jit

        res = run_jit(fn=main, specs=_build_specs(), compile_only=True,
                      runtime_cfg=dict(platform=args.platform, device_id=args.device))
        print(f"[merge_rope_standalone] compile-only passed={res.passed} error={res.error!r}")
    else:
        main_run(device=args.device, platform=args.platform)
```
</details>

## Questions for PTO-ISA

1. Is there a Vec UB region PTOAS reserves (pipe-init/validate, or runtime scratch) that a kernel's `alloc_tile` allocations must **not** cross? What is the **real usable Vec UB** for `a2a3` / `Ascend910B1` on cann-9.0.0?
2. We were told the reservation was removed — but it still corrupts at 188.8 KB on 9.0.0. Is it still in effect, or is this a separate near-full-footprint bug?
3. If it's a fixed reservation, should the upstream allocator subtract it from the 192 KB budget so an over-tall layout **errors at compile** instead of corrupting silently? Can you confirm the exact reserved size?
